### PR TITLE
Mat conversions for macOS/AppKit

### PR DESCRIPTION
**Pull request by mtfrctl**
This PR solves https://github.com/opencv/opencv/issues/18546.
It consists

- Extract CoreGraphics conversion logics from ios_conversions.mm to apple_conversions.h, apple_conversions.mm (MatToCGImage(), CGImageToMat())
- Simplify ios_conversions.mm
- Add macosx_conversions.mm for macOS/AppKit (MatToNSImage(), NSImageToMat())
- Add Mat+Conversions.h and Mat+Conversions.mm for macOS/AppKit

<cut/>

**Usage example in Swift:**
**※ Don't forget to set the `-all_load` flag to "Other Linker Flags" in "Build Settings" when using!** https://github.com/opencv/opencv/issues/17532
```
// CGImage to Mat
let sourceMat = Mat(cgImage: cgImage, alphaExist: true)

// NSImage to Mat
let sourceMat = Mat(nsImage: nsImage, alphaExist: true)

// Something to process
let grayMat = Mat()
Imgproc.cvtColor(src: sourceMat, dst: grayMat, code: .COLOR_RGBA2GRAY)

// Mat to CGImage
let cgImage = grayMat.toCGImage().takeRetainedValue() // or .takeUnetainedValue()

// Mat to NSImage
let nsImage = grayMat.toNSImage()
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work

```
force_builders=Custom Mac
build_image:Custom Mac=osx_framework

buildworker:iOS=macosx-1
```

### CODE REVIEW

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+//
+//  IMPORTANT: READ BEFORE DOWNLOADING, COPYING, INSTALLING OR USING.
+//
+//  By downloading, copying, installing or using the software you agree to this license.
+//  If you do not agree to this license, do not download, install,
+//  copy or use the software.
+//
+//
+//                          License Agreement
+//                For Open Source Computer Vision Library
```
Comment (alalek) [inline modules/imgcodecs/include/opencv2/imgcodecs/macosx.h:12]: Please use short license header: https://github.com/opencv/opencv/wiki/Coding_Style_Guide#file-structure

```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
```

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
```
Comment (alalek) [inline modules/imgcodecs/include/opencv2/imgcodecs/macosx.h:43]: It make sense to add some check here with proper error message:

```
#if !defined(__APPLE__)
#error This header should be used in ObjC/Swift projects.
```

(condition may be improved, just a sample)

File: modules/imgcodecs/src/apple_conversions.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#import <Accelerate/Accelerate.h>
+#import <AVFoundation/AVFoundation.h>
+#import <ImageIO/ImageIO.h>
+#include "opencv2/core.hpp"
+#include "precomp.hpp"
```
Comment (alalek) [inline modules/imgcodecs/src/apple_conversions.h:47]: > precomp.hpp

Please don't use this from header files. Move into .mm files instead

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#import <AppKit/AppKit.h>
+#import <Accelerate/Accelerate.h>
+#import <AVFoundation/AVFoundation.h>
+#import <ImageIO/ImageIO.h>
```
Comment (alalek) [inline modules/imgcodecs/include/opencv2/imgcodecs/macosx.h:47]: > Accelerate

Is it necessary? What code uses that?

File: modules/imgcodecs/src/apple_conversions.mm [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+#include "apple_conversions.h"
+
+CGImageRef MatToCGImage(const cv::Mat& image) {
+    NSData *data = [NSData dataWithBytes:image.data
+                                  length:image.step.p[0] * image.rows];
+
+    CGColorSpaceRef colorSpace;
+
+    if (image.elemSize() == 1) {
```
Comment (alalek) [inline modules/imgcodecs/src/apple_conversions.mm:51]: `.elemSize()` => `.channels()`

\+ `CV_CheckDepth(image.depth(), CV_8U, "");` check on the first line.

---

OK, I see this code is just moved.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

Please take a look on comments below.

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
+
```
Comment (alalek) [inline modules/imgcodecs/include/opencv2/imgcodecs/macosx.h:1]: To resolve problem with ABI checker on Linux builder please add this platform-specific header into skip list [here](https://github.com/opencv/opencv/blob/76be3529f491aa5835602021c65ef99975837cac/cmake/templates/opencv_abi.xml.in#L34).

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h (hunk 1 above)
Comment (mtfrctl) [reply modules/imgcodecs/include/opencv2/imgcodecs/macosx.h]: 🙆‍♂️

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h (hunk 2 above)
Comment (mtfrctl) [reply modules/imgcodecs/include/opencv2/imgcodecs/macosx.h:9]: Thank you. I added a compile guard for non-macOS.

File: modules/imgcodecs/src/apple_conversions.h (hunk 3 above)
Comment (mtfrctl) [reply modules/imgcodecs/src/apple_conversions.h]: Thank you. I moved it.

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h (hunk 4 above)
Comment (mtfrctl) [reply modules/imgcodecs/include/opencv2/imgcodecs/macosx.h]: I didn't need it after all, so I removed it. Thank you.

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h (hunk 6 above)
Comment (mtfrctl) [reply modules/imgcodecs/include/opencv2/imgcodecs/macosx.h:1]: I was having trouble getting rid of the ABI error, but I overlooked this file. Thank you for pointing it out! 🙇‍♂️

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h (hunk 1 above)
Comment (alalek) [reply modules/imgcodecs/include/opencv2/imgcodecs/macosx.h]: Thank you for updates!
Please make similar update of license headers of new files in this PR.
Please remove empty first line in this file.

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h (hunk 2 above)
Comment (mtfrctl) [reply modules/imgcodecs/include/opencv2/imgcodecs/macosx.h:8]: Fixed guard condition && -> ||
https://github.com/opencv/opencv/pull/18547/commits/7b668e460e9bfcfa22628d2c8cd88c2dd965a1fc

File: modules/imgcodecs/include/opencv2/imgcodecs/macosx.h (hunk 1 above)
Comment (mtfrctl) [reply modules/imgcodecs/include/opencv2/imgcodecs/macosx.h]: https://github.com/opencv/opencv/pull/18547/commits/ae0ba082e222870f3a850c115c874e36575b321e

**Comment by komakai** [comment]
@mtfrctl thanks for this - this is a great feature.
I checked the code and everything looks fine.
I have tried it quickly - after I remembered to set **-all_load** in the **Other Linker Flags** of my project everything worked well.
👍

**Comment by mtfrctl** [comment]
@komakai Thank you! Regarding the flag, people who come to check here will also check these @komakai and @treastrain conversations (https://github.com/opencv/opencv/issues/17532 https://github.com/opencv/opencv/issues/18023), but it would be nice in the future so that it can be used immediately without being aware of these things.

**Comment by mtfrctl** [comment]
@alalek By the way, I would like to add the following method to `modules/core/misc/objc/commonMat.mm (and .h)`, is it okay to include it in this PR, or is it better to separate it into another PR?
```
- (unsigned char*)dataPtr {
    return _nativePtr->data;
}
```

**Comment by alalek** [comment]
If there is no strong dependency between patches, then separate PR is better.

**Comment by mtfrctl** [comment]
@alalek I created a PR separately here https://github.com/opencv/opencv/pull/18566, so I'd be happy if you could check it 🙏

**Review by alalek — APPROVED** [review]
@mtfrctl Thank you for contribution :+1:

@komakai Thank you for review!

**Comment by komakai** [comment]
@mtfrctl keep the contributions coming!

**Pull request merged**